browser/src/control/Control.AutoFillPreviewTooltip.ts

```diff
@@ -0,0 +1,129 @@
+/* -*- js-indent-level: 8 -*- */
+/*
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+/*
+ * Control.AutoFillPreviewTooltip - class for tooltip of cell previews during auto fill
+ */
+
+/* global app */
+
+class AutoFillPreviewTooltip extends L.Control.AutoCompletePopup {
+	usageText: string;
+	newPopupData: PopupData;
+
+	constructor(map: ReturnType<typeof L.map>) {
+		super('autoFillPreviewTooltip', map);
+		this.newPopupData = {
+			children: [
+				{
+					id: 'container',
+					type: 'container',
+					enabled: true,
+					children: new Array<WidgetJSON>(),
+					vertical: true,
+				} as any as WidgetJSON,
+			] as Array<WidgetJSON>,
+			jsontype: 'dialog',
+			type: 'modalpopup',
+			cancellable: false,
+			popupParent: '_POPOVER_',
+			clickToClose: '_POPOVER_',
+			id: 'autoFillPreviewTooltip',
+			title: 'HELLO',
+		} as PopupData;
+	}
+
+	onAdd() {
+		this.newPopupData.isAutoCompletePopup = true;
+		this.map.on('openautofillpopup', this.openAutoFillPopup, this);
+		this.map.on('closementionpopup', this.closeMentionPopup, this);
+		this.map.on('sendautofilllocation', this.sendAutoFillLocation, this);
+		// this.map.on('mousedown', this.onMouseDown, this);
+		// this.map.on('mouseup', this._onMouseUp, this);
+	}
+
+	// private onMouseUp (e: MouseEvent) {
+	//     console.error("yaz: " + e);
+	// }
+
+	getAutoFillAreaBottomRightLocation(ev: FireEvent): Point {
+		// var currPos = ev.data.location;
+		var currPos = {
+			// x: app.file.textCursor.rectangle.cX1,
+			x: app.file.viewedRectangle.cX1,
+			y: app.file.textCursor.rectangle.cY2,
+		};
+		// console.error("currPos: " + currPos);
+		var origin = this.map.getPixelOrigin();
+		var panePos = this.map._getMapPanePos();
+		// this._currMousePos = { x: ev.originalEvent.pageX, y: ev.originalEvent.pageY };
+		return new L.Point(
+			Math.round(currPos.x + panePos.x - origin.x),
+			Math.round(currPos.y + panePos.y - origin.y),
+		);
+	}
+
+	sendAutoFillLocation(ev: FireEvent) {
+		// console.error("data: " + ev.data.location);
+		this.openAutoFillPopup({ data: ev.data });
+	}
+
+	getPopupEntries(ev: FireEvent): any[] {
+		const entries: any[] = [{ row: '0', columns: [{ text: ev.data.text }] }];
+		return entries;
+	}
+
+	openAutoFillPopup(ev: FireEvent): void {
```

Comment:
isn't that something we can get from the base class AutoCompletePopup? or do we have some changes here?

Reply:
yes, we use `getSimpleTextJSON(entry);` here. So we need to modify `openAutoFillPopup` function here.

Reply:
but the position calculation differes than the base one?

Reply:
Yes, position calculation on the base one is following the cursor while editing the cell. But we need to get the bottom right location of the cell selected with autofill.